api/grpcserver/v2beta1/smeshing_identities_test.go

```diff
@@ -0,0 +1,193 @@
+package v2beta1
+
+import (
+	"context"
+	"maps"
+	"math"
+	"slices"
+	"testing"
+	"time"
+
+	pb "github.com/spacemeshos/api/release/go/spacemesh/v2beta1"
+	"github.com/stretchr/testify/require"
+	"go.uber.org/zap/zaptest"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+	"google.golang.org/protobuf/types/known/timestamppb"
+
+	"github.com/spacemeshos/go-spacemesh/activation"
+	"github.com/spacemeshos/go-spacemesh/common/types"
+	"github.com/spacemeshos/go-spacemesh/identity"
+	"github.com/spacemeshos/go-spacemesh/sql/localsql"
+)
+
+func TestSmeshingIdentitiesService_States(t *testing.T) {
+	t.Run("request validation", func(t *testing.T) {
+		states := identity.NewIdentityStateStorage(localsql.InMemoryTest(t), zaptest.NewLogger(t))
+		svc := NewSmeshingIdentitiesService(states, nil, activation.PoetConfig{})
+		tests := []struct {
+			name    string
+			req     *pb.IdentityStatesRequest
+			wantErr codes.Code
+		}{
+			{
+				name:    "limit exceeds maximum",
+				req:     &pb.IdentityStatesRequest{Limit: 101},
+				wantErr: codes.InvalidArgument,
+			},
+			{
+				name:    "zero limit",
+				req:     &pb.IdentityStatesRequest{Limit: 0},
+				wantErr: codes.InvalidArgument,
+			},
+			{
+				name: "invalid from timestamp",
+				req: &pb.IdentityStatesRequest{
+					Limit: 10,
+					From:  &timestamppb.Timestamp{Seconds: math.MinInt64},
+				},
+				wantErr: codes.InvalidArgument,
+			},
+			{
+				name: "invalid to timestamp",
+				req: &pb.IdentityStatesRequest{
+					Limit: 10,
+					To:    &timestamppb.Timestamp{Seconds: math.MaxInt64},
+				},
+				wantErr: codes.InvalidArgument,
+			},
+		}
+
+		for _, tc := range tests {
+			t.Run(tc.name, func(t *testing.T) {
+				_, err := svc.States(context.Background(), tc.req)
+				st, ok := status.FromError(err)
+				require.True(t, ok)
+				require.Equal(t, tc.wantErr, st.Code())
+			})
+		}
+	})
+}
+
+func TestSmeshingIdentitiesService_StatesFiltering(t *testing.T) {
+	statesDB := identity.NewIdentityStateStorage(localsql.InMemoryTest(t), zaptest.NewLogger(t))
+	svc := NewSmeshingIdentitiesService(statesDB, nil, activation.PoetConfig{})
+
+	nodeID := types.RandomNodeID()
+
+	firstTimestamp := time.Now().Add(-time.Minute)
+	lastTimestamp := firstTimestamp
+	nextTimestamp := func() time.Time {
+		t := lastTimestamp
+		lastTimestamp = lastTimestamp.Add(time.Second)
+		return t
+	}
+
+	broadcasted := identity.ATXBroadcasted{
+		AtxId:   types.RandomATXID(),
+		Publish: 9,
+	}
+	states := map[time.Time]identity.State{
+		nextTimestamp(): &identity.ATXReady{
+			Publish: 9,
+		},
+		nextTimestamp(): &broadcasted,
+		nextTimestamp(): &identity.WaitForATXSynced{},
+		nextTimestamp(): &identity.WaitForPoetRoundEnd{},
+	}
```

Comment:
NIT: why not keeping it simple? This is shorter and imo clearer (`lastTimeStamp` isn't even used)

```suggestion
	broadcasted := identity.ATXBroadcasted{
		AtxId:   types.RandomATXID(),
		Publish: 9,
	}

	firstTimeStamp := time.Now().Add(-time.Minute)
	states := map[time.Time]identity.State{
		firstTimeStamp.Add(1 * time.Second): &identity.ATXReady{
			Publish: 9,
		},
		firstTimeStamp.Add(2 * time.Second): &broadcasted,
		firstTimeStamp.Add(3 * time.Second): &identity.WaitForATXSynced{},
		firstTimeStamp.Add(4 * time.Second): &identity.WaitForPoetRoundEnd{},
	}
	// lastTimeStamp := max(maps.Keys(states)) // not actually used anywhere
```